hack/update: base library

prezha opened the pull request:
fixes: #9464 

pr #9464 ("automate update processes") split: part 1/4 - this one should be the first one to merge

includes:
 - update.go: package init, base struct w/ method, and functions
 - filesystem.go: update local filesystem repo
 - github.go: get greatest stable and latest releases, find/create PR,
   and update remote GitHub repo
 - registry.go: manage images (eg, pull, tag, etc.) and update container
   registries

<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->

medyagh reviewed (changes requested):
thank you for splitting the work into smaller parts
do you mind choosing a more declaring name, for both the PR and also the folder, so it is obvious what is the "update" for

prezha commented:
@medyagh no worries
also, please note that this part is still wip - i'll update when it's ready (in the next few hours) so that it's not reviewed multiple times...

not sure how to rename the update folder (and the pr), as it will contain more declaring subfolders and files like:
- part 2 pr would be named `hack-update--kubernetes-version` and will contain script: `hack/update/kubernetes_version/update_kubernetes_version.go`,
- part 3 pr would be named `hack-update--golang-version` and will contain script: `hack/update/golang_version/update_golang_version.go`, and
- part 4 pr would be named `hack-update--kicbase-version` and will contain script: `hack/update/kicbase_version/update_kicbase_version.go`

does that makes sense or we need to have something more specific names? (if latter - please help with proposals ! :))

prezha commented:
this one should be ok to merge now

medyagh commented:
PR needs lint

medyagh reviewed (changes requested):
plz check lint

prezha commented:
> plz check lint

@medyagh, yes, the lint error here was expected - please see:

https://github.com/kubernetes/minikube/pull/9526#issuecomment-718265171 above:
> this one should be ok to merge now (lint error `hack/update/kubernetes_version/update_kubernetes_version.go:76:32: GHVersions not declared by package update` is covered in pr #9529)

and

https://github.com/kubernetes/minikube/pull/9529#issuecomment-718264649:
> @tstromberg similar as with pr #9532:
> 
> the conflicts were due to the pr #9464 (the original one - before i've split it into several prs) got merged first, and i had a couple of changes made in those individual prs; essentially my fault, i should have had closed the original pr right after the split (but left it just in case anyone wanted to compare)...
> 
> i've fixed this one now but had to also add here the `base library` from pr #9526 to resolve dependencies, should be ok to merge this as well as pr #9526 (i'll resolve the conflicts there if any)

we can:
- merge this one and then also with merging #9529 everything should be fine,
or
- merge #9529 first and then this one (i'll resolve any conflicts)

or if you have a better proposal - please share your thoughts

prezha commented:
this one got automatically closed after merging #9529 and rebasing, as explained in https://github.com/kubernetes/minikube/pull/9529#issuecomment-718264649
